File: Templates/Day_4/Orphaning_Nowait/Fortran/vectorNormModularSerial.f90
Program VectorNormModular

  use vectorNormSizes
  
  implicit none

  double precision, dimension(:), allocatable :: vect
  double precision :: norm

  integer :: iloop
  
  print *, "Length:", vleng

  allocate(vect(vleng))
  
  ! measure multiple times to stabilise timing results - ncycles is defined in vectorNormSizes.f90
  ! unccomment the loop when required
  ! do iloop = 1, ncycles

  call vectorInit(vect, vleng)
  norm = vectorNormSqr(vect,vleng)
  norm = sqrt(norm)
  ! enddo ! loop over iloop

  ! print result
  print *,"Norm::", norm
  
  deallocate(vect)

contains

  subroutine vectorInit(v, leng)
    integer, intent(in)  :: leng
    double precision, dimension(leng), intent(out) :: v

    integer i

    ! orphan directive
    ! default static schedule is required in connection with the nowait statement
    !$omp do schedule(static)
    do i=1, leng
       v(i) = dble(i)
    enddo
    !$omp end do nowait

  end subroutine vectorInit

  !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
  
  function vectorNormSqr(v, leng)
    integer, intent(in) :: leng
    double precision, dimension(leng), intent(in) :: v
    double precision :: vectorNormSqr

    integer i

    vectorNormSqr = 0.0d0
    ! orphan directive
    ! default static schedule is required in connection with the nowait statement
    !$omp do schedule(static)
    do i=1, leng
       vectorNormSqr = VectorNormSqr + v(i) * v(i)
    enddo
    !$omp end do nowait

  end function vectorNormSqr
  
  
End Program VectorNormModular
  

File: Templates/Day_4/Orphaning_Nowait/C/vectorNormModularSerial.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

#include "vectorNormSizes.h"

void vectorInit(double* v, int leng);
double vectorNormSqr(double* v, int leng);

int main()
{

  printf("length: %i\n", vleng);
  double *vect; 
  double norm;

  vect = malloc(vleng * sizeof(double));

  // measure multiple times to stabilise timing results - cycles is defined in vectorNormSizes.h
  // uncomment the loop when required
  //for (int iloop=0; iloop < cycles; iloop++)
    {      
      
      vectorInit(vect, vleng);	
      norm = vectorNormSqr(vect, vleng);	
      norm = sqrt(norm);
    }

  printf("Norm: %f\n",norm);

  free(vect);
  
}



void vectorInit(double* v, int leng)
{
  for (int i=0; i < leng; i++)
    v[i] = (double) (i+1); 
  return;
}

double vectorNormSqr(double* v, int leng)
{
  double norm = 0.0;
  for (int i=0; i < leng; i++)
    {
      norm += v[i] * v[i];
    }

  return norm;
}
